diff --git a/fontforge/charview.c b/fontforge/charview.c
index 242a6db..54f6dfe 100644
--- a/fontforge/charview.c
+++ b/fontforge/charview.c
@@ -70,6 +70,8 @@ float arrowAccelFactor=10.;
 float snapdistance=3.5;
 int updateflex = false;
 extern int clear_tt_instructions_when_needed;
+int use_freetype_with_aa_fill_cv = 1;
+int prefs_preview_mode_uses_cairo_fill = 0;
 int interpCPsOnMotion=false;
 int default_cv_width = 540;
 int default_cv_height = 540;
@@ -105,7 +107,8 @@ struct cvshows CVShows = {
 	0,		/* show curves which are almost, but not exactly horizontal or vertical at the end-points */
 	3,		/* number of em-units a coord difference must be less than to qualify for almost hv */
 	1,		/* Check for self-intersections in the element view */
-	1		/* In tt debugging, mark changed rasters differently */
+	1,		/* In tt debugging, mark changed rasters differently */
+	0               /* show filled by directly drawing the fill using cairo */
 };
 struct cvshows CVShowsPrevewToggleSavedState;
 
@@ -1062,10 +1065,16 @@ return;
 }
 
 void CVDrawSplineSet(CharView *cv, GWindow pixmap, SplinePointList *set,
-	Color fg, int dopoints, DRect *clip ) {
+		     Color fg, int dopoints, DRect *clip ) {
+    CVDrawSplineSetSpecialized( cv, pixmap, set, fg, dopoints, clip, sfm_stroke );
+}
+
+void CVDrawSplineSetSpecialized(CharView *cv, GWindow pixmap, SplinePointList *set,
+				Color fg, int dopoints, DRect *clip, enum outlinesfm_flags strokeFillMode ) {
     Spline *spline, *first;
     SplinePointList *spl;
     int truetype_markup = set==cv->b.gridfit && cv->dv!=NULL;
+    int currentSplineCounter = 0;
 
     if ( cv->inactive )
 	dopoints = false;
@@ -1104,9 +1113,14 @@ void CVDrawSplineSet(CharView *cv, GWindow pixmap, SplinePointList *set,
 	    x = rpt(cv,  cv->xoff + spl->first->me.x*cv->scale);
 	    y = rpt(cv, -cv->yoff + cv->height - spl->first->me.y*cv->scale);
 	    GDrawPathMoveTo(pixmap,x+.5,y+.5);
+	    currentSplineCounter++;
 	    for ( spline=spl->first->next, first=NULL; spline!=first && spline!=NULL; spline=spline->to->next ) {
 		x = rpt(cv,  cv->xoff + spline->to->me.x*cv->scale);
 		y = rpt(cv, -cv->yoff + cv->height - spline->to->me.y*cv->scale);
+		/* printf("linear:%d order2:%d else:%d\n", */
+		/*        spline->knownlinear, */
+		/*        spline->order2, */
+		/*        !(spline->knownlinear || spline->order2)); */
 		if ( spline->knownlinear )
 		    GDrawPathLineTo(pixmap,x+.5,y+.5);
 		else if ( spline->order2 ) {
@@ -1139,7 +1153,25 @@ void CVDrawSplineSet(CharView *cv, GWindow pixmap, SplinePointList *set,
 	    }
 	    if ( spline!=NULL )
 		GDrawPathClose(pixmap);
-	    GDrawPathStroke(pixmap,(spl->is_clip_path ? clippathcol : fg)|0xff000000);
+	    printf("pathfill splintCounter:%d sfm:%d sf:%d sfc:%d spl:%p\n",
+		   currentSplineCounter, strokeFillMode,
+		   cv->showfilled, cv->showfilledusingcairo,
+		   spl );
+
+	    switch( strokeFillMode )
+	    {
+	    case sfm_stroke:
+		GDrawPathStroke(pixmap,(spl->is_clip_path ? clippathcol : fg)|0xff000000);
+		break;
+	    case sfm_fill:
+		if( currentSplineCounter == 1 ) {
+		    GDrawPathFill( pixmap, (spl->is_clip_path ? clippathcol : fg)|0xff000000);
+		} else {
+		    GDrawPathFill( pixmap, (default_background)|0xff000000);
+		}
+		break;
+	    }
+	    
 	} else {
 	    GPointList *gpl = MakePoly(cv,spl), *cur;
 	    for ( cur=gpl; cur!=NULL; cur=cur->next )
@@ -1154,7 +1186,7 @@ void CVDrawSplineSet(CharView *cv, GWindow pixmap, SplinePointList *set,
 }
 
 static void CVDrawLayerSplineSet(CharView *cv, GWindow pixmap, Layer *layer,
-	Color fg, int dopoints, DRect *clip ) {
+				 Color fg, int dopoints, DRect *clip, enum outlinesfm_flags strokeFillMode ) {
 #ifdef FONTFORGE_CONFIG_TYPE3
     int active = cv->b.layerheads[cv->b.drawmode]==layer;
     int ml = cv->b.sc->parent->multilayer;
@@ -1183,7 +1215,7 @@ static void CVDrawLayerSplineSet(CharView *cv, GWindow pixmap, Layer *layer,
 	GDrawSetLineWidth(pixmap,0);
 #endif
 #else
-    CVDrawSplineSet(cv,pixmap,layer->splines,fg,dopoints,clip);
+    CVDrawSplineSetSpecialized(cv,pixmap,layer->splines,fg,dopoints,clip,strokeFillMode);
 #endif
 }
 
@@ -1935,7 +1967,7 @@ static void CVDrawGridRaster(CharView *cv, GWindow pixmap, DRect *clip ) {
     }
     if ( cv->showback[0]&1 ) {
 	CVDrawSplineSet(cv,pixmap,cv->b.gridfit,gridfitoutlinecol,
-		cv->showpoints,clip);
+			cv->showpoints,clip);
     }
 }
 
@@ -2194,6 +2226,28 @@ return;				/* no points. no side bearings */
     }
 }
 
+static int CVExposeGlyphFill(CharView *cv, GWindow pixmap, GEvent *event, DRect* clip ) {
+    int i, layer, rlayer, cvlayer = CVLayer((CharViewBase *) cv);
+    int filled = 0;
+    if (( cv->showfore || cv->b.drawmode==dm_fore ) && cv->showfilled &&
+	cv->filled!=NULL ) {
+	GDrawDrawImage(pixmap, &cv->gi, NULL,
+		       cv->xoff + cv->filled->xmin,
+		       -cv->yoff + cv->height-cv->filled->ymax);
+	filled = 1;
+    }
+
+    if( cv->showfilledusingcairo ) {
+	layer = cvlayer;
+	if ( layer>=0 ) {
+	    CVDrawLayerSplineSet(cv,pixmap,&cv->b.sc->layers[layer],foreoutlinecol,
+				 cv->showpoints, clip, sfm_fill );
+	    filled = 1;
+	}
+    }
+    return(filled);
+}
+
 static void CVExpose(CharView *cv, GWindow pixmap, GEvent *event ) {
     SplineFont *sf = cv->b.sc->parent;
     RefChar *rf;
@@ -2202,9 +2256,14 @@ static void CVExpose(CharView *cv, GWindow pixmap, GEvent *event ) {
     char buf[20];
     PST *pst;
     int i, layer, rlayer, cvlayer = CVLayer((CharViewBase *) cv);
-
+    enum outlinesfm_flags strokeFillMode = sfm_stroke;
+    int GlyphHasBeenFilled = 0;
     GDrawPushClip(pixmap,&event->u.expose.rect,&old);
 
+    if( cv->showfilledusingcairo==1 ) {
+	strokeFillMode = sfm_fill;
+    }
+    
     clip.width = event->u.expose.rect.width/cv->scale;
     clip.height = event->u.expose.rect.height/cv->scale;
     clip.x = (event->u.expose.rect.x-cv->xoff)/cv->scale;
@@ -2275,13 +2334,9 @@ static void CVExpose(CharView *cv, GWindow pixmap, GEvent *event ) {
 
 	DrawSelImageList(cv,pixmap,cv->b.layerheads[cv->b.drawmode]->images);
 
-	if (( cv->showfore || cv->b.drawmode==dm_fore ) && cv->showfilled &&
-		cv->filled!=NULL ) {
-	    /* Wrong order, I know. But it is useful to have the background */
-	    /*  visible on top of the fill... */
-	    GDrawDrawImage(pixmap, &cv->gi, NULL, cv->xoff + cv->filled->xmin,
-		    -cv->yoff + cv->height-cv->filled->ymax);
-	}
+	/* Wrong order, I know. But it is useful to have the background */
+	/*  visible on top of the fill... */
+	GlyphHasBeenFilled = CVExposeGlyphFill(cv, pixmap, event, &clip );
     } else {
 	/* Draw FreeType Results */
 	CVDrawGridRaster(cv,pixmap,&clip);
@@ -2303,8 +2358,8 @@ static void CVExpose(CharView *cv, GWindow pixmap, GEvent *event ) {
 	    /* Then when we moved the fill image around, we had to deal with the */
 	    /*  images before the fill... */
 	    CVDrawLayerSplineSet(cv,pixmap,&cv->b.sc->layers[layer],
-		    !sf->multilayer || layer==ly_back ? backoutlinecol : foreoutlinecol,
-		    false,&clip);
+				 !sf->multilayer || layer==ly_back ? backoutlinecol : foreoutlinecol,
+				 false,&clip,false);
 	    for ( rf=cv->b.sc->layers[layer].refs; rf!=NULL; rf = rf->next ) {
 		if ( /* cv->b.drawmode==dm_back &&*/ cv->showrefnames )
 		    CVDrawRefName(cv,pixmap,rf,0);
@@ -2327,11 +2382,24 @@ static void CVExpose(CharView *cv, GWindow pixmap, GEvent *event ) {
     /*  draw the current layer -- unless they've turned on grid fit. Then they*/
     /*  might want to hide the active layer. */
     layer = cvlayer;
+
+    /* The call to CVExposeGlyphFill() above will have rendered a filled glyph already. */
+    /* We draw the outline only at this stage so as to have it layered */
+    /* over the control points if they are currently visible. */
+    /* CVDrawLayerSplineSet() will draw both the control points, and the font outline over those */
+    if( cv->showfilledusingcairo==1 ) {
+	strokeFillMode = sfm_stroke;
+    }
+    if( GlyphHasBeenFilled ) {
+//	strokeFillMode = sfm_nothing;
+	strokeFillMode = sfm_stroke;
+    }
+    
     if ( layer<0 ) /* Guide lines are special */
 	CVDrawLayerSplineSet(cv,pixmap,cv->b.layerheads[cv->b.drawmode],foreoutlinecol,
-		cv->showpoints ,&clip);
+			     cv->showpoints ,&clip,strokeFillMode);
     else if ( (cv->showback[layer>>5]&(1<<(layer&31))) ||
-	    (!cv->show_ft_results && cv->dv==NULL )) {
+	      (!cv->show_ft_results && cv->dv==NULL )) {
 	for ( rf=cv->b.sc->layers[layer].refs; rf!=NULL; rf = rf->next ) {
 	    if ( cv->showrefnames )
 		CVDrawRefName(cv,pixmap,rf,0);
@@ -2340,10 +2408,9 @@ static void CVExpose(CharView *cv, GWindow pixmap, GEvent *event ) {
 	    if ( rf->selected && cv->b.layerheads[cv->b.drawmode]==&cv->b.sc->layers[layer])
 		CVDrawBB(cv,pixmap,&rf->bb);
 	}
-
-	CVDrawLayerSplineSet(cv,pixmap,&cv->b.sc->layers[layer],foreoutlinecol,
-		cv->showpoints ,&clip);
     }
+    CVDrawLayerSplineSet(cv,pixmap,&cv->b.sc->layers[layer],foreoutlinecol,
+			 cv->showpoints ,&clip,strokeFillMode);
 
     if ( cv->freehand.current_trace!=NULL )
 	CVDrawSplineSet(cv,pixmap,cv->freehand.current_trace,tracecol,
@@ -2450,18 +2517,27 @@ void CVRegenFill(CharView *cv) {
 	/*  fill. But on the "M" (and "W") glyph of extravigant caps, ft won't*/
 	/*  do a mono fill */
 	if ( use_freetype_to_rasterize_fv && hasFreeType()) {
+	    int depth = 1;
+	    if( use_freetype_with_aa_fill_cv ) {
+		depth = 4;
+		clut_len = 16;
+	    }
+	    printf("aa fill. use_freetype_with_aa_fill_cv:%d depth:%d\n",use_freetype_with_aa_fill_cv,depth);
 	    cv->filled = SplineCharFreeTypeRasterizeNoHints(cv->b.sc,layer,
-		size,72, 1);
+		size,72, depth);
 	    if ( cv->filled==NULL && size<2000 ) {
 		/* There are some glyphs which freetype won't rasterize in */
 		/* mono mode, but will in grey scale. Don't ask me why */
+		printf("aa fill. going grey on second attempt\n");
 		cv->filled = SplineCharFreeTypeRasterizeNoHints(cv->b.sc,
 		    layer, size, 72, 4);
 		clut_len = 16;
 	    }
 	}
-	if ( cv->filled==NULL )
+	if ( cv->filled==NULL ) {
+	    printf("spline rasterize....\n");
 	    cv->filled = SplineCharRasterize(cv->b.sc,layer,size+.1);
+	}
 	if ( cv->filled==NULL )
 return;
 	cv->gi.u.image->image_type = clut_len==2 ? it_mono : it_index;
@@ -5047,6 +5123,7 @@ return( true );
 #define MID_Prev	2008
 #define MID_HideRulers	2009
 #define MID_Preview     2010
+#define MID_FillUsingCairo 2011
 #define MID_NextDef	2012
 #define MID_PrevDef	2013
 #define MID_DisplayCompositions	2014
@@ -5750,10 +5827,21 @@ static void CVMenuFill(GWindow gw,struct gmenuitem *mi,GEvent *e) {
     CharView *cv = (CharView *) GDrawGetUserData(gw);
 
     CVShows.showfilled = cv->showfilled = !cv->showfilled;
+    CVShows.showfilledusingcairo = cv->showfilledusingcairo = 0;
     CVRegenFill(cv);
     GDrawRequestExpose(cv->v,NULL,false);
 }
 
+static void CVMenuFillUsingCairo(GWindow gw,struct gmenuitem *mi,GEvent *e) {
+    CharView *cv = (CharView *) GDrawGetUserData(gw);
+
+    CVShows.showfilledusingcairo = cv->showfilledusingcairo = !cv->showfilledusingcairo;
+    CVShows.showfilled = cv->showfilled = 0;
+    CVRegenFill(cv);
+    GDrawRequestExpose(cv->v,NULL,false);
+}
+
+
 static struct cvshows* cvshowsCopyTo( struct cvshows* dst, CharView* src ) 
 {
     dst->showfore = src->showfore;
@@ -5763,6 +5851,7 @@ static struct cvshows* cvshowsCopyTo( struct cvshows* dst, CharView* src )
     dst->showdhints = src->showdhints;
     dst->showpoints = src->showpoints;
     dst->showfilled = src->showfilled;
+    dst->showfilledusingcairo = src->showfilledusingcairo;
     dst->showrulers = src->showrulers;
     dst->showrounds = src->showrounds;
     dst->showmdx = src->showmdx;
@@ -5795,6 +5884,7 @@ static CharView* cvshowsCopyFrom( CharView* dst, struct cvshows* src )
     dst->showdhints = src->showdhints;
     dst->showpoints = src->showpoints;
     dst->showfilled = src->showfilled;
+    dst->showfilledusingcairo = src->showfilledusingcairo;
     dst->showrulers = src->showrulers;
     dst->showrounds = src->showrounds;
     dst->showmdx = src->showmdx;
@@ -5833,7 +5923,8 @@ void CVPreviewModeSet(GWindow gw, int checked ) {
         cv->showvhints = 0;
         cv->showdhints = 0;
         cv->showpoints = 0;
-        cv->showfilled = 1;
+        cv->showfilled = !prefs_preview_mode_uses_cairo_fill;
+	cv->showfilledusingcairo = prefs_preview_mode_uses_cairo_fill;
         cv->showrounds = 0;
         cv->showanchor = 0;
         cv->showrefnames = 0;
@@ -9261,6 +9352,9 @@ static void swlistcheck(GWindow gw,struct gmenuitem *mi,GEvent *e) {
 	  case MID_Fill:
 	    mi->ti.checked = cv->showfilled;
 	  break;
+	  case MID_FillUsingCairo:
+	    mi->ti.checked = cv->showfilledusingcairo;
+	  break;
 	  case MID_ShowHHints:
 	    mi->ti.checked = cv->showhhints;
 	    mi->ti.disabled = sf->multilayer;
@@ -10010,7 +10104,8 @@ static GMenuItem2 swlist[] = {
     { { (unichar_t *) N_("(Define \"Almost\")"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 0, 0, 0, 0, 1, 1, 0, 'M' }, H_("(Define \"Almost\")|No Shortcut"), NULL, NULL, CVMenuDefineAlmost, MID_DefineAlmost },
     { { (unichar_t *) N_("_Side Bearings"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'M' }, H_("Side Bearings|No Shortcut"), NULL, NULL, CVMenuShowSideBearings, MID_ShowSideBearings },
     { { (unichar_t *) N_("Reference Names"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'M' }, H_("Reference Names|No Shortcut"), NULL, NULL, CVMenuShowRefNames, MID_ShowRefNames },
-    { { (unichar_t *) N_("_Fill"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'l' }, H_("Fill|No Shortcut"), NULL, NULL, CVMenuFill, MID_Fill },
+    { { (unichar_t *) N_("_Fill Using Freetype"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'l' }, H_("Fill|No Shortcut"), NULL, NULL, CVMenuFill, MID_Fill },
+    { { (unichar_t *) N_("Fil_l Using Cairo"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'l' }, H_("Fill|No Shortcut"), NULL, NULL, CVMenuFillUsingCairo, MID_FillUsingCairo },
     { { (unichar_t *) N_("Previe_w"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'l' }, H_("Preview|Ctl+`"), NULL, NULL, CVMenuPreview, MID_Preview },
     { { NULL, NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 0, 0, 0, 1, 0, 0, 0, '\0' }, NULL, NULL, NULL, NULL, 0 }, /* line */
     { { (unichar_t *) N_("Pale_ttes"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 0, 0, 0, 0, 1, 1, 0, 'P' }, NULL, pllist, pllistcheck, NULL, 0 },
@@ -10278,6 +10373,7 @@ static void _CharViewCreate(CharView *cv, SplineChar *sc, FontView *fv,int enc)
     cv->showpoints = CVShows.showpoints;
     cv->showrulers = CVShows.showrulers;
     cv->showfilled = CVShows.showfilled;
+    cv->showfilledusingcairo = CVShows.showfilledusingcairo;
     cv->showrounds = CVShows.showrounds;
     cv->showmdx = CVShows.showmdx;
     cv->showmdy = CVShows.showmdy;
diff --git a/fontforge/prefs.c b/fontforge/prefs.c
index badbc54..39a16d9 100644
--- a/fontforge/prefs.c
+++ b/fontforge/prefs.c
@@ -55,6 +55,8 @@ extern int adjustlbearing;
 extern Encoding *default_encoding;
 extern int autohint_before_generate;
 extern int use_freetype_to_rasterize_fv;
+extern int use_freetype_with_aa_fill_cv;
+extern int prefs_preview_mode_uses_cairo_fill;
 extern int OpenCharsInNewWindow;
 extern int ItalicConstrained;
 extern int accent_offset;
@@ -279,9 +281,11 @@ static struct prefs_list {
 	{ N_("HelpDir"), pr_file, &helpdir, NULL, NULL, 'H', NULL, 0, N_("The directory on your local system in which FontForge will search for help\nfiles.  If a file is not found there, then FontForge will look for it on the net.") },
 	{ N_("OtherSubrsFile"), pr_file, &othersubrsfile, NULL, NULL, 'O', NULL, 0, N_("If you wish to replace Adobe's OtherSubrs array (for Type1 fonts)\nwith an array of your own, set this to point to a file containing\na list of up to 14 PostScript subroutines. Each subroutine must\nbe preceded by a line starting with '%%%%' (any text before the\nfirst '%%%%' line will be treated as an initial copyright notice).\nThe first three subroutines are for flex hints, the next for hint\nsubstitution (this MUST be present), the 14th (or 13 as the\nnumbering actually starts with 0) is for counter hints.\nThe subroutines should not be enclosed in a [ ] pair.") },
 	{ N_("FreeTypeInFontView"), pr_bool, &use_freetype_to_rasterize_fv, NULL, NULL, 'O', NULL, 0, N_("Use the FreeType rasterizer (when available)\nto rasterize glyphs in the font view.\nThis generally results in better quality.") },
+	{ N_("FreeTypeAAFillInOutlineView"), pr_bool, &use_freetype_with_aa_fill_cv, NULL, NULL, 'O', NULL, 0, N_("When filling using freetype in the outline view,\nhave freetype render the glyph antialiased.") },
 	{ N_("SplashScreen"), pr_bool, &splash, NULL, NULL, 'S', NULL, 0, N_("Show splash screen on start-up") },
 #ifndef _NO_LIBCAIRO
 	{ N_("UseCairoDrawing"), pr_bool, &prefs_usecairo, NULL, NULL, '\0', NULL, 0, N_("Use the cairo library for drawing (if available)\nThis makes for prettier (anti-aliased) but slower drawing\nThis applies to any windows created AFTER this is set.\nAlready existing windows will continue as they are.") },
+	{ N_("PreviewModeFillCairo"), pr_bool, &prefs_preview_mode_uses_cairo_fill, NULL, NULL, 'O', NULL, 0, N_("When in Preview Mode in the outline view,\nuse cairo to render the glyph fill.") },
 #endif
 #ifndef _NO_LIBPANGO
 	{ N_("UsePangoDrawing"), pr_bool, &prefs_usepango, NULL, NULL, '\0', NULL, 0, N_("Use the pango library for text (if available)\nThis makes for prettier and handles complex scripts.\nBut it can slow things down on older machines.\nThis applies to any windows created AFTER this is set.\nAlready existing windows will continue as they are.") },
@@ -468,6 +472,7 @@ static struct prefs_list {
 	{ "AnchorControlPixelSize", pr_int, &aa_pixelsize, NULL, NULL, '\0', NULL, 1, NULL },
 #ifdef _NO_LIBCAIRO
 	{ "UseCairoDrawing", pr_bool, &prefs_usecairo, NULL, NULL, '\0', NULL, 0, N_("Use the cairo library for drawing (if available)\nThis makes for prettier (anti-aliased) but slower drawing\nThis applies to any windows created AFTER this is set.\nAlready existing windows will continue as they are.") },
+	{ N_("PreviewModeFillCairo"), pr_bool, &prefs_preview_mode_uses_cairo_fill, NULL, NULL, 'O', NULL, 0, N_("When in Preview Mode in the outline view,\nuse cairo to render the glyph fill.") },
 #endif
 #ifdef _NO_LIBPANGO
 	{ "UsePangoDrawing", pr_bool, &prefs_usepango, NULL, NULL, '\0', NULL, 0, N_("Use the cairo library for drawing (if available)\nThis makes for prettier (anti-aliased) but slower drawing\nThis applies to any windows created AFTER this is set.\nAlready existing windows will continue as they are.") },
diff --git a/fontforge/views.h b/fontforge/views.h
index cac0cf4..b98fa5c 100644
--- a/fontforge/views.h
+++ b/fontforge/views.h
@@ -56,6 +56,7 @@ extern struct cvshows {
     int hvoffset;
     int checkselfintersects;	/* Not really something shown, but convenient to keep it here */
     int showdebugchanges;	/* Changes the way changing rasters are displayed in tt debug mode */
+    int showfilledusingcairo;   /* fill the glyph, but directly use cairo to do the fill rather than freetype */
 } CVShows;
 
 extern struct bvshows {
@@ -165,6 +166,7 @@ typedef struct charview {
     unsigned int checkselfintersects: 1;
     unsigned int showdebugchanges: 1;
     unsigned int inPreviewMode: 1;
+    unsigned int showfilledusingcairo:1;
     int hvoffset;		/* for showalmosthvlines */
     int layers_off_top;
     real scale;
@@ -786,8 +788,11 @@ extern void SFPrivateInfo(SplineFont *sf);
 extern void FVDelay(FontView *fv,void (*func)(FontView *));
 extern void GFI_FinishContextNew(struct gfi_data *d,FPST *fpst, int success);
 extern void SCPreparePopup(GWindow gw,SplineChar *sc, struct remap *remap, int enc, int actualuni);
+enum outlinesfm_flags { sfm_stroke=0x1, sfm_fill=0x2, sfm_nothing=0x4 };
+extern void CVDrawSplineSetSpecialized(CharView *cv, GWindow pixmap, SplinePointList *set,
+				       Color fg, int dopoints, DRect *clip, enum outlinesfm_flags strokeFillMode );
 extern void CVDrawSplineSet(CharView *cv, GWindow pixmap, SplinePointList *set,
-	Color fg, int dopoints, DRect *clip );
+			    Color fg, int dopoints, DRect *clip );
 extern GWindow CVMakeTools(CharView *cv);
 extern GWindow CVMakeLayers(CharView *cv);
 extern GWindow BVMakeTools(BitmapView *bv);
