diff --git a/fontforge/charview.c b/fontforge/charview.c
index afa6e47..decdb20 100644
--- a/fontforge/charview.c
+++ b/fontforge/charview.c
@@ -107,6 +107,11 @@ struct cvshows CVShows = {
 	1,		/* Check for self-intersections in the element view */
 	1		/* In tt debugging, mark changed rasters differently */
 };
+struct cvshows CVShowsPrevewToggleSavedState;
+
+#define CID_Base	1001
+
+
 static Color pointcol = 0xff0000;
 static Color firstpointcol = 0x707000;
 static Color selectedpointcol = 0xc8c800;
@@ -3020,6 +3025,10 @@ static void CVDoFindInFontView(CharView *cv) {
     GDrawRaise(((FontView *) (cv->b.fv))->gw);
 }
 
+static uint16 HaveModifiers = 0;
+static uint16 PressingTilde = 0;
+static uint16 PrevCharEventWasCharUpOnControl = 0;
+
 static void CVCharUp(CharView *cv, GEvent *event ) {
 
     #if defined(__MINGW32__)
@@ -3027,6 +3036,55 @@ static void CVCharUp(CharView *cv, GEvent *event ) {
 	_mingw_hand_tool_hack(cv);
     #endif
 
+    if( event->u.chr.keysym=='`' ) 
+    {
+        PressingTilde = 1;
+    }
+
+    if( PrevCharEventWasCharUpOnControl
+        && event->u.chr.keysym=='`' )
+    {
+        HaveModifiers = 0;
+        PrevCharEventWasCharUpOnControl = 0;
+        return;
+    }
+    
+    PrevCharEventWasCharUpOnControl = 0;
+    
+    if( !event->u.chr.autorepeat
+        && (event->u.chr.keysym == GK_Control_L
+            || event->u.chr.keysym == GK_Control_R ))
+    {
+        PrevCharEventWasCharUpOnControl = 1;
+        if( !PressingTilde )
+        {
+            printf( "charup() ... control released and not pressing tilde\n");
+            HaveModifiers = 0;
+        }
+    }
+
+    printf( "charup()   repeat:%d havemod:%d press`:%d lctl:%d keysym:%c\n",
+            event->u.chr.autorepeat,
+            HaveModifiers,
+            PressingTilde,
+            (event->u.chr.keysym == GK_Control_L),
+            event->u.chr.keysym );
+    if ( !event->u.chr.autorepeat && !HaveModifiers && event->u.chr.keysym=='`' ) 
+    {
+        PressingTilde = 0;
+        CVPreviewModeSet( cv->gw, false );
+        return;
+    }
+    if ( !event->u.chr.autorepeat && event->u.chr.keysym=='`' )
+    {
+        PressingTilde = 0;
+    }
+    if ( event->u.chr.autorepeat && HaveModifiers && event->u.chr.keysym=='`' )
+    {
+        return;
+    }
+    
+    
 #if _ModKeysAutoRepeat
     /* Under cygwin these keys auto repeat, they don't under normal X */
     if ( event->u.chr.keysym == GK_Shift_L || event->u.chr.keysym == GK_Shift_R ||
@@ -5195,6 +5254,9 @@ return( true );
 
 #define MID_Warnings	3000
 
+#define MID_Preview     4000
+
+
 static void CVMenuClose(GWindow gw,struct gmenuitem *mi,GEvent *e) {
     CharView *cv = (CharView *) GDrawGetUserData(gw);
     if ( cv->b.container )
@@ -5706,6 +5768,116 @@ static void CVMenuFill(GWindow gw,struct gmenuitem *mi,GEvent *e) {
     GDrawRequestExpose(cv->v,NULL,false);
 }
 
+static struct cvshows* cvshowsCopyTo( struct cvshows* dst, CharView* src ) 
+{
+    dst->showfore = src->showfore;
+//    dst->showback = src->showback;
+    dst->showgrids = src->showgrids;
+    dst->showhhints = src->showhhints;
+    dst->showvhints = src->showvhints;
+    dst->showdhints = src->showdhints;
+    dst->showpoints = src->showpoints;
+    dst->showfilled = src->showfilled;
+    dst->showrulers = src->showrulers;
+    dst->showrounds = src->showrounds;
+    dst->showmdx = src->showmdx;
+    dst->showmdy = src->showmdy;
+    dst->showhmetrics = src->showhmetrics;
+    dst->showvmetrics = src->showvmetrics;
+    dst->markextrema = src->markextrema;
+    dst->markpoi = src->markpoi;
+    dst->showblues = src->showblues;
+    dst->showfamilyblues = src->showfamilyblues;
+    dst->showanchor = src->showanchor;
+    dst->showcpinfo = src->showcpinfo;
+    dst->showtabs = src->showtabs;
+    dst->showsidebearings = src->showsidebearings;
+    dst->showrefnames = src->showrefnames;
+    dst->snapoutlines = src->snapoutlines;
+    dst->showalmosthvlines = src->showalmosthvlines;
+    dst->showalmosthvcurves = src->showalmosthvcurves;
+    dst->hvoffset = src->hvoffset;
+    dst->checkselfintersects = src->checkselfintersects;
+    dst->showdebugchanges = src->showdebugchanges;
+    return dst;
+}
+static CharView* cvshowsCopyFrom( CharView* dst, struct cvshows* src ) 
+{
+    dst->showfore = src->showfore;
+//    dst->showback = src->showback;
+    dst->showgrids = src->showgrids;
+    dst->showhhints = src->showhhints;
+    dst->showvhints = src->showvhints;
+    dst->showdhints = src->showdhints;
+    dst->showpoints = src->showpoints;
+    dst->showfilled = src->showfilled;
+    dst->showrulers = src->showrulers;
+    dst->showrounds = src->showrounds;
+    dst->showmdx = src->showmdx;
+    dst->showmdy = src->showmdy;
+    dst->showhmetrics = src->showhmetrics;
+    dst->showvmetrics = src->showvmetrics;
+    dst->markextrema = src->markextrema;
+    dst->markpoi = src->markpoi;
+    dst->showblues = src->showblues;
+    dst->showfamilyblues = src->showfamilyblues;
+    dst->showanchor = src->showanchor;
+    dst->showcpinfo = src->showcpinfo;
+    dst->showtabs = src->showtabs;
+    dst->showsidebearings = src->showsidebearings;
+    dst->showrefnames = src->showrefnames;
+    dst->snapoutlines = src->snapoutlines;
+    dst->showalmosthvlines = src->showalmosthvlines;
+    dst->showalmosthvcurves = src->showalmosthvcurves;
+    dst->hvoffset = src->hvoffset;
+    dst->checkselfintersects = src->checkselfintersects;
+    dst->showdebugchanges = src->showdebugchanges;
+    return dst;
+}
+
+void CVPreviewModeSet(GWindow gw, int checked ) {
+    CharView *cv = (CharView *) GDrawGetUserData(gw);
+    if( checked && cv->inPreviewMode )
+        return;
+
+    cv->inPreviewMode = checked;
+    if( checked )
+    {
+        cvshowsCopyTo( &CVShowsPrevewToggleSavedState, cv );
+        cv->showfore   = 1;
+        cv->showgrids  = 0;
+        cv->showhhints = 0;
+        cv->showvhints = 0;
+        cv->showdhints = 0;
+        cv->showpoints = 0;
+        cv->showfilled = 1;
+        cv->showrounds = 0;
+        cv->showanchor = 0;
+        cv->showrefnames = 0;
+        cv->showhmetrics = 0;
+        cv->showvmetrics = 0;
+    }
+    else
+    {
+        cvshowsCopyFrom( cv, &CVShowsPrevewToggleSavedState );
+    }
+    CVRegenFill(cv);
+    GDrawRequestExpose(cv->v,NULL,false);
+}
+
+
+                                    
+static void CVMenuPreview(GWindow gw,struct gmenuitem *mi,GEvent *e) {
+    CharView *cv = (CharView *) GDrawGetUserData(gw);
+    int checked = mi->ti.checked;
+
+    printf("CVMenuPreview() HaveModifiers:%d presstilde:%d e:%p\n", HaveModifiers, PressingTilde, e );
+    if( !HaveModifiers )
+        return;
+    
+    CVPreviewModeSet( gw, checked );
+}
+
 static void CVMenuShowGridFit(GWindow gw,struct gmenuitem *mi,GEvent *e) {
     CharView *cv = (CharView *) GDrawGetUserData(gw);
 
@@ -5943,10 +6115,32 @@ static void CVMenuChangeChar(GWindow gw,struct gmenuitem *mi,GEvent *e) {
     _CVMenuChangeChar(cv,mi->mid);
 }
 
+
 void CVChar(CharView *cv, GEvent *event ) {
     extern float arrowAmount, arrowAccelFactor;
     extern int navigation_mask;
 
+   
+    if( event->u.chr.keysym == GK_Control_L
+        || event->u.chr.keysym == GK_Control_R )
+    {
+        HaveModifiers = 1;
+    }
+
+    printf( "chardown() repeat:%d havemod:%d press`:%d lctl:%d keysym:%c\n",
+            event->u.chr.autorepeat,
+            HaveModifiers,
+            PressingTilde,
+            (event->u.chr.keysym == GK_Control_L),
+            event->u.chr.keysym );
+    
+    if( !HaveModifiers && event->u.chr.keysym=='`' ) 
+    {
+        PressingTilde = 1;
+        CVPreviewModeSet( cv->gw, true );
+        return;
+    }
+    
 #if _ModKeysAutoRepeat
 	/* Under cygwin these keys auto repeat, they don't under normal X */
 	if ( cv->autorpt!=NULL ) {
@@ -9841,6 +10269,7 @@ static GMenuItem2 swlist[] = {
     { { (unichar_t *) N_("_Side Bearings"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'M' }, H_("Side Bearings|No Shortcut"), NULL, NULL, CVMenuShowSideBearings, MID_ShowSideBearings },
     { { (unichar_t *) N_("Reference Names"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'M' }, H_("Reference Names|No Shortcut"), NULL, NULL, CVMenuShowRefNames, MID_ShowRefNames },
     { { (unichar_t *) N_("_Fill"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'l' }, H_("Fill|No Shortcut"), NULL, NULL, CVMenuFill, MID_Fill },
+    { { (unichar_t *) N_("Previe_w"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'l' }, H_("Preview|Ctl+`"), NULL, NULL, CVMenuPreview, MID_Preview },
     { { NULL, NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 0, 0, 0, 1, }},
     { { (unichar_t *) N_("Pale_ttes"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 0, 0, 0, 0, 1, 1, 0, 'P' }, NULL, pllist, pllistcheck },
     { { (unichar_t *) N_("_Glyph Tabs"), NULL, COLOR_DEFAULT, COLOR_DEFAULT, NULL, NULL, 0, 1, 1, 0, 0, 0, 1, 1, 0, 'R' }, H_("Glyph Tabs|No Shortcut"), NULL, NULL, CVMenuShowTabs, MID_ShowTabs },
@@ -10125,7 +10554,7 @@ static void _CharViewCreate(CharView *cv, SplineChar *sc, FontView *fv,int enc)
     cv->showanchor = CVShows.showanchor;
     cv->showcpinfo = CVShows.showcpinfo;
     cv->showtabs = CVShows.showtabs;
-
+    cv->inPreviewMode = 0;
     cv->checkselfintersects = CVShows.checkselfintersects;
 
     cv->showdebugchanges = CVShows.showdebugchanges;
diff --git a/fontforge/views.h b/fontforge/views.h
index b726ccf..cac0cf4 100644
--- a/fontforge/views.h
+++ b/fontforge/views.h
@@ -164,6 +164,7 @@ typedef struct charview {
     unsigned int showalmosthvcurves: 1;
     unsigned int checkselfintersects: 1;
     unsigned int showdebugchanges: 1;
+    unsigned int inPreviewMode: 1;
     int hvoffset;		/* for showalmosthvlines */
     int layers_off_top;
     real scale;
diff --git a/gdraw/gxdraw.c b/gdraw/gxdraw.c
index d409ca1..bb74562 100644
--- a/gdraw/gxdraw.c
+++ b/gdraw/gxdraw.c
@@ -3471,6 +3471,7 @@ return;
 	gevent.type = event->type==KeyPress?et_char:et_charup;
 	gevent.u.chr.time = event->xkey.time;
 	gevent.u.chr.state = event->xkey.state;
+            gevent.u.chr.autorepeat = 0;
 /*#ifdef __Mac*/
 	/* On mac os x, map the command key to the control key. So Comand-Q=>^Q=>Quit */
 	/* No... don't. Let the user have access to the command key as distinct from control */
@@ -3559,6 +3560,26 @@ return;
 	    gevent.u.chr.keysym = keysym;
 	    gevent.u.chr.chars[0] = '\0';
 	}
+
+            /*
+             * If we are a charup, but the very next XEvent is a chardown
+             * on the same key, then we are just an autorepeat XEvent which
+             * other code might like to ignore
+             */
+	if ( gevent.type==et_charup && XEventsQueued(gdisp->display, QueuedAfterReading))
+            {
+                XEvent nev;
+                XPeekEvent(gdisp->display, &nev);
+                if (nev.type == KeyPress && nev.xkey.time == event->xkey.time &&
+                    nev.xkey.keycode == event->xkey.keycode)
+                {
+                    gevent.u.chr.autorepeat = 1;
+                }
+                printf( "autorepeat:%d us.code:%ld next.code:%ld\n",
+                        gevent.u.chr.autorepeat,
+                        event->xkey.keycode,
+                        nev.xkey.keycode );
+            }
       break;
       case ButtonPress: case ButtonRelease: case MotionNotify:
 	if ( expecting_core && gdisp->last_event_time==event->xbutton.time )
diff --git a/inc/gdraw.h b/inc/gdraw.h
index 7eedc67..2a6c1e5 100644
--- a/inc/gdraw.h
+++ b/inc/gdraw.h
@@ -141,6 +141,7 @@ typedef struct gevent {
 	    uint16 state;
 	    int16 x,y;
 	    uint16 keysym;
+        int16 autorepeat;
 	    unichar_t chars[_GD_EVT_CHRLEN];
 	} chr;
 	struct {
